### PR TITLE
AI map v2

### PR DESCRIPTION
**Pull request by TobleroneSwordfish**
[FEATURE] [EXPERIMENTAL]
## About the PR 
A complete rewrite of #6779 implemented in a hopefully less cursed way. This one builds a colour coded minimap as an icon at roundstart, then displays it as a map component in a tgui window.
kondaru
Other maps: https://imgur.com/a/CVeXD0c
Some of the map code is loosely based off StationMap.dm, but I chose to rewrite it in a different file rather than try to adapt that unfinished code.

Known issues:
- The first time the map is opened it takes around 10 seconds to render correctly. This isn't actual loading time and I'm not really sure what causes it.
- Some of the colour areas look a little weird because there are some tiles with the wrong areas set (like the infuriating single red pixel in Kondaru botany)

## Why's this needed? 
People seemed to really like the idea of this, but the last PR was built on pure jank. This one is hopefully better.

## Changelog

```changelog
(u)LeahTheTech
(*)Added a map for the AI to move around with. Open with the "Open station map" verb and click to teleport.
```

### CODE REVIEW

**Comment by virvatuli-gh** [comment]
I can't comment/review on the code aspect of this but as someone who plays AI a lot, I really dig the concept. It'd be a great QoL feature especially for newer players who aren't as familiar with station layouts who wind up as AI.

**Comment by mordent-goonstation** [comment]
Given that this map is generated at roundstart, could these be static images that are created in response to changes to the station .dmm files instead of during round initialization? Not entirely sure what that process would look like, but that happens a lot less than roundstarts, for sure.

**Comment by TobleroneSwordfish** [comment]
> could these be static images that are created in response to changes to the station .dmm files instead of during round initialization?

They could be, but that would involve either automagically starting the game and running inputs to generate the images or parsing dmm files directly (which I have no idea how to do, and can't find any documentation on).
In any case, the actual generation seems to take less than a second every roundstart on my local tests and unless I've really fucked up it shouldn't be any different on live.
I'll try to get an exact time for generating Cog2 actually.

**Comment by TobleroneSwordfish** [comment]
It seems to reliably take 2 ticks to generate (0.2 seconds), so I think the benefits of being able to just parse the game world directly outweigh the extra loading time.

**Comment by ZeWaka** [comment]
> It seems to reliably take 2 ticks to generate (0.2 seconds), so I think the benefits of being able to just parse the game world directly outweigh the extra loading time.

It'll be even faster on our game servers, for reference. I'd expect one tick.

File: code/mob/living/intangible/ai-camera.dm [hunk 1]
```diff
@@ -37,6 +37,8 @@
 
 	var/outer_eye_atom = null
 
+	var/datum/tgui/map_ui
```
Comment (ZeWaka) [inline code/mob/living/intangible/ai-camera.dm:40]: code comment on this pls

File: code/global.dm [hunk 2]
```diff
@@ -506,6 +506,8 @@ var/global
 
 	hardRebootFilePath = "data/hard-reboot"
 
+	obj/station_map/ai_station_map
```
Comment (ZeWaka) [inline code/global.dm:509]: code comment on this pls

File: goonstation.dme [hunk 3]
```diff
@@ -1871,6 +1871,7 @@ var/datum/preMapLoad/preMapLoad = new
 #include "code\z_dmm_suite\map_placeholders.dm"
 #include "code\z_dmm_suite\reader.dm"
 #include "code\z_dmm_suite\writer.dm"
+#include "icons\obj\station_map.dm"
```
Comment (ZeWaka) [inline goonstation.dme:1874]: no code inside the icons folder. please mode this to a folder in `modules/`

File: icons/obj/station_map.dm [hunk 4]
```diff
@@ -0,0 +1,89 @@
+///This is exclusively for the AI map right now
+/obj/station_map
+	name = "AI station map"
+	var/static/icon/map_icon
+
+	var/x_max = 1
+	var/x_min
+	var/y_max = 1
+	var/y_min
```
Comment (ZeWaka) [inline icons/obj/station_map.dm:9]:
```suggestion
	var/x_min = null
	var/y_max = 1
	var/y_min = null
```
More clarity that you set it in the constructor and thought about this.

File: icons/obj/station_map.dm [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+		icon = map_icon
+
+	Click(location, control, params)
+		if (!isAIeye(usr)) //no humans teleporting
+			return
+		var/list/param_list = params2list(params)
+		if ("left" in param_list)
+			var/x = text2num(param_list["icon-x"])
+			var/y = text2num(param_list["icon-y"])
+			var/turf/clicked = locate(x, y, Z_LEVEL_STATION) //station z level macro
```
Comment (ZeWaka) [inline icons/obj/station_map.dm:32]:
```suggestion
			var/turf/clicked = locate(x, y, Z_LEVEL_STATION)
```
unneces.

File: icons/obj/station_map.dm [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+			usr.loc = clicked
+		if ("right" in param_list)
+			return TRUE
+
+	//generates the map from the current station layout
+	proc/render_map()
+		for (var/y = world.maxy, y > 0, y--)
```
Comment (ZeWaka) [inline icons/obj/station_map.dm:39]: Unsure if you're unaware of fast fors with a backwards step or you have a reason.
```suggestion
		for (var/y in world.maxy to 1 step -1)
```

**Review by ZeWaka — CHANGES_REQUESTED** [review]

File: icons/obj/station_map.dm (hunk 6 above)
Comment (TobleroneSwordfish) [reply icons/obj/station_map.dm:39]: yep, did not know the step syntax (couldn't see it in the byond docs anywhere?)

File: goonstation.dme (hunk 3 above)
Comment (TobleroneSwordfish) [reply goonstation.dme:1874]: ha

File: code/global.dm (hunk 2 above)
Comment (ZeWaka) [reply code/global.dm:510]: code doc comment

File: code/mob/living/intangible/ai-camera.dm (hunk 1 above)
Comment (ZeWaka) [reply code/mob/living/intangible/ai-camera.dm:40]: code doc comment

Comment (TobleroneSwordfish) [reply code/mob/living/intangible/ai-camera.dm:40]: oh of course, sorry

**Comment by mordent-goonstation** [comment]
> It seems to reliably take 2 ticks to generate (0.2 seconds), so I think the benefits of being able to just parse the game world directly outweigh the extra loading time.

Works for me.

**Comment by TobleroneSwordfish** [comment]
Well instead of just not working the first time it opens, it now only doesn't work for around 10 seconds the first time it opens.

**Review by ZeWaka — APPROVED** [review]

**Review by pali6 — APPROVED** [review]
rad

**Pull request merged**